pipenv/cli/options.py

```diff
@@ -0,0 +1,385 @@
+# -*- coding=utf-8 -*-
+from __future__ import absolute_import
+
+import os
+
+from click import BOOL as click_booltype
+from click import (
+    BadParameter, Group, Option, argument, echo, make_pass_decorator, option
+)
+
+from .. import environments
+from ..utils import is_valid_url
+
+
+CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])
+
+
+class PipenvGroup(Group):
+    """Custom Group class provides formatted main help"""
+
+    def get_help_option(self, ctx):
+        from ..core import format_help
+
+        """Override for showing formatted main help via --help and -h options"""
+        help_options = self.get_help_option_names(ctx)
+        if not help_options or not self.add_help_option:
+            return
+
+        def show_help(ctx, param, value):
+            if value and not ctx.resilient_parsing:
+                if not ctx.invoked_subcommand:
+                    # legit main help
+                    echo(format_help(ctx.get_help()))
+                else:
+                    # legit sub-command help
+                    echo(ctx.get_help(), color=ctx.color)
+                ctx.exit()
+
+        return Option(
+            help_options,
+            is_flag=True,
+            is_eager=True,
+            expose_value=False,
+            callback=show_help,
+            help="Show this message and exit.",
+        )
+
+
+class State(object):
+    def __init__(self):
+        self.index = None
+        self.extra_index_urls = []
+        self.verbose = False
+        self.pypi_mirror = None
+        self.python = None
+        self.two = None
+        self.three = None
+        self.site_packages = False
+        self.clear = False
+        self.system = False
+        self.installstate = InstallState()
+
+
+class InstallState(object):
+    def __init__(self):
+        self.dev = False
+        self.pre = False
+        self.selective_upgrade = False
+        self.keep_outdated = False
+        self.skip_lock = False
+        self.ignore_pipfile = False
+        self.sequential = False
+        self.code = False
+        self.requirementstxt = None
+        self.deploy = False
+        self.packages = []
+        self.editables = []
+
+
+pass_state = make_pass_decorator(State, ensure=True)
+
+
+def index_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.index = value
+        return value
+    return option('-i', '--index', expose_value=False, envvar="PIP_INDEX_URL",
+                        help='Target PyPI-compatible package index url.', nargs=1,
+                        callback=callback)(f)
+
+
+def extra_index_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        if isinstance(value, (tuple, list)):
+            state.extra_index_urls.extend(list(value))
```

Comment:
No need to convert this one extra time, `extend()` works for any iterables. But is this check (to distinguish input type from strings) needed at all? Doesn’t the `multiple=True` flag guarantee this always gets a list?

Reply:
Hmmm I think so, will fix